### PR TITLE
Reactify project header

### PR DESCRIPTION
**Pull request by Hamms**
Our current implementation of the various ways to render headers for Project views are all implemented in jquery. Specifically, we have a handful of methods on `dashboard.header` that are called by `dashboard.project` to initialize different kinds of headers for different situations, and to update the components in response to other lifecycle events.

This reactification keeps the `dashboard.header` methods around, but changes them to simply be Redux proxies. All the rendering logic now lives in a bunch of new React components.

### CODE REVIEW

**Review by islemaster — COMMENTED** [review]
This is awesome and I really want to do it.  Also, my eyes are glazing over a little bit trying to review this - there's a lot of subtle logic being moved over into the new header components, and I don't trust myself to catch problems at the moment.  Do you have a recommended approach to review? Would I be better off walking through the commit history here?

Also, I don't see any stories or tests for the new components.  I understand that our UI and eyes tests should provide fairly good coverage of this behavior, but this seems like a great opportunity to add unit test coverage, especially for things like the logic buried in the `ProjectRemix` button.

**Comment by Hamms** [comment]
Haha, yeah, it's a bit much. The commit history should be a good place to start, particularly the first few commits where I'm mostly just extracting things into React. Another potential way to structure your review is to work through the header methods one at a time and compare them to the React replacements.

Note also that the logic for the most complicated things, like the Remix and Import buttons, was simply lifted entire from header.js rather than written specifically for this PR.

As Dave pointed out, https://github.com/code-dot-org/code-dot-org/pull/28421 is an isolated PR containing just the first commit of this one. If this PR is too much to review as a monolith, I could try to break it up into incremental PRs in the spirit of that one.

Tests are forthcoming, in a separate PR! I just figured this one was getting to be too big as it is.

**Comment by islemaster** [comment]
Thanks, that helps.  I don't object to the large PR, as long as you'll help us through it.

IMO the primary purpose of small PRs is as a forcing function on cohesion and ratchet-ability. This change exhibits both properties already, and in some ways it's much nicer to see the whole plan in one place rather than broken up.

I do have a personal preference for seeing tests go in side-by-side with changes like this, but will defer to your judgment.

**Comment by davidsbailey** [comment]
@Hamms I've taken a look at #28421 and that seems mergeable after addressing comments. https://github.com/code-dot-org/code-dot-org/pull/28422 (containing the next PR of this commit) seems like a reasonable thing to merge next. I'll add some review comments there.

**Comment by davidsbailey** [comment]
just to add a bit to the earlier discussion, I'm not super comfortable with the merge-first-and-test-later approach here because this is a huge refactor affecting live code. I would normally only encourage that pattern when writing a new feature that is either not live or protected by an experiment flag. if we think that the existing UI test coverage actually has us well covered here then that's a different line of reasoning which I am open to hearing.

regarding PR size, while I think perhaps the first 2-3 commits could all go out in a single deploy, sending this all out at once seems like it would create quite a large surface area for new problems to turn up in the deploy (which is a likely way we'd discover problems given the limited test coverage) and that could make it much more difficult to isolate which part of the monolithic PR introduced the problem.

thirdly, I forgot to say at the beginning that I am in fact very excited for this change! the header code has been a nagging area for quite some time and I'm very happy to see it tackled!

**Comment by Hamms** [comment]
> if we think that the existing UI test coverage actually has us well covered here then that's a different line of reasoning which I am open to hearing.

Oh, sorry I hadn't made that explicit. Yes, the only reason I'm comfortable merging the unit tests later is because our existing UI tests cover header functionality quite extensively. That being said, if that approach is making everyone else uncomfortable then I'm more than happy to include the unit tests in this PR.

**Comment by Hamms** [comment]
@islemaster @davidsbailey @joshlory PTAL; I merged some of the commits from this PR as standalone changes, and added unit tests

File: apps/src/code-studio/components/header/ProjectHeader.jsx
Comment (islemaster) [inline apps/src/code-studio/components/header/ProjectHeader.jsx:55]: Will you tell me more about this? I _think _ our UI tests interface with controlled components okay in other places.  If not, that's something we should solve (but probably not in this PR).

File: apps/src/code-studio/components/header/ProjectHeader.jsx
Comment (islemaster) [inline apps/src/code-studio/components/header/ProjectHeader.jsx:113]: :joy: So... never?

File: apps/src/code-studio/components/header/ProjectInfo.jsx
Comment (islemaster) [inline apps/src/code-studio/components/header/ProjectInfo.jsx:17]: I really enjoyed this elegant way to make the inner components mutually exclusive.

File: apps/src/code-studio/components/header/ProjectRemix.jsx
Comment (islemaster) [inline apps/src/code-studio/components/header/ProjectRemix.jsx:25]: This was true of the original code too, but this seems like a good time to ask: Why don't we have to urlencode the `user_return_to` parameter here?

File: apps/src/code-studio/components/header/ProjectRemix.jsx
Comment (islemaster) [inline apps/src/code-studio/components/header/ProjectRemix.jsx]: New TODO.

File: apps/src/code-studio/components/header/ProjectUpdatedAt.jsx
Comment (islemaster) [inline apps/src/code-studio/components/header/ProjectUpdatedAt.jsx]: Another TODO.

**Review by islemaster — APPROVED** [review]
Changes look great! I have some questions but everything looks correct.  What do you see as next steps in this area? Is there further simplification you're hoping to do?

File: apps/src/code-studio/components/header/ProjectRemix.jsx
Comment (davidsbailey) [reply apps/src/code-studio/components/header/ProjectRemix.jsx]: +1, I believe TODOs should ideally be removed, but if not then should at least have your name on them.

File: apps/src/code-studio/components/header/ProjectHeader.jsx
Comment (davidsbailey) [reply apps/src/code-studio/components/header/ProjectHeader.jsx:55]: I do know that there are various combinations of `I press keys` and `I type "..." into` and possibly others which work in some situations and not others, presumably based on which particular events get fired by each. would be great if we had one UI test step that triggered all the right events... but for now could be useful to name which steps you tried that didn't work with a controlled component.

File: apps/src/code-studio/components/header/ProjectHeader.jsx
Comment (davidsbailey) [inline apps/src/code-studio/components/header/ProjectHeader.jsx:76]: historically we (especially Brent) pushed hard to not have control flow logic mixed into rendering logic, fact we've largely avoided these kind of renderXyz methods entirely. I believe the goal is for the rendering code to look as much like the document it renders as possible. How does this sound as a compromise:
```
  renderSaveOrEdit() {
    this.state.editName ? renderSave() : renderEdit()
```
and then, I suppose it would be up to you whether to inline the now-very-simple `renderSaveOrEdit` below.

The more extreme suggestion would be to extract new react components (which in this case could be simple [stateless functional components](https://hackernoon.com/react-stateless-functional-components-nine-wins-you-might-have-overlooked-997b0d933dbc) defined within this file) when the logic grows dense enough that subcomponents need to be extracted.

@Hamms and @islemaster , open to pushback on whether these kinds of things are something we care to push on moving forward.

File: apps/src/code-studio/components/header/ProjectImport.jsx
Comment (davidsbailey) [inline apps/src/code-studio/components/header/ProjectImport.jsx:18]: yuck! of course this yuckiness has to go somewhere and is a necessary evil for now given what you're doing, but I really don't like seeing it inside a react component. Would this seem any less gross if moved to headerRedux?

to be really clean, the Craft class would trigger a redux action which sets an importProject callback, so that all the Craft code can stay inside craft. but that seems like too much for now.

ok if you decide not to do anything here right now and def don't think we need to track this explicitly if we don't do it now, but when badness appears it can be a good prompt to at least identify a few steps along the path to making it better.

File: apps/src/code-studio/components/header/ProjectInfo.jsx
Comment (davidsbailey) [reply apps/src/code-studio/components/header/ProjectInfo.jsx:17]: +1! Agree that life would be worse if you had to use inline control flow exclusively here.

**Review by davidsbailey — COMMENTED** [review]
LGTM after addressing Brad's concerns. Also, I should add that this change feels very big to review in one go, so I'm relying on the assumption that we have good UI test coverage here to make sure this is working end-to-end. Thank you for tackling this! Getting the header onto React and Redux is a huge improvement overall and seems totally work any risk of a small regression.

File: apps/src/code-studio/components/header/ProjectHeader.jsx
Comment (Hamms) [reply apps/src/code-studio/components/header/ProjectHeader.jsx:55]: The specific tests I found that failed in the controlled input scenario were using the `type_into_selector` helper: https://github.com/code-dot-org/code-dot-org/blob/02a1bfbc7f4580ea36aa219150d34f473379b9b3/dashboard/test/ui/features/step_definitions/steps.rb#L549-L554

Which uses `$.change()`, which doesn't trigger the change event in a way that can be detected by React. It's possible that it could be modified to do so, but React is picky about only picking up user-triggered events and not code-triggered events, and it seemed easier to just make the component uncontrolled.

File: apps/src/code-studio/components/header/ProjectHeader.jsx
Comment (Hamms) [reply apps/src/code-studio/components/header/ProjectHeader.jsx:113]: :upside_down_face:

File: apps/src/code-studio/components/header/ProjectInfo.jsx
Comment (Hamms) [reply apps/src/code-studio/components/header/ProjectInfo.jsx:17]: Thanks! I was actually halfway through implementing an inline control flow version when I realized this would be _way_ easier

File: apps/src/code-studio/components/header/ProjectRemix.jsx
Comment (Hamms) [reply apps/src/code-studio/components/header/ProjectRemix.jsx:25]: Is there a built-in guarantee that `window.location.pathname` is already urlencoded? It seems reasonable that there would be

File: apps/src/code-studio/components/header/ProjectRemix.jsx
Comment (Hamms) [reply apps/src/code-studio/components/header/ProjectRemix.jsx]: I'll replace these TODOs with some jira cards

**Comment by Hamms** [comment]
> What do you see as next steps in this area?

In the short term, I'd like to assemble a comprehensive list of all the functional components in the header and where and how they're all defined, then continue to convert them into React. I think the sign in button and the create dropdown are probably good candidates for next targets. Ultimately, once everything is all defined in React I'd like to unify the entire header into a single set of components.

> I should add that this change feels very big to review in one go, so I'm relying on the assumption that we have good UI test coverage here to make sure this is working end-to-end

Likewise :P

File: apps/src/code-studio/components/header/ProjectHeader.jsx
Comment (Hamms) [reply apps/src/code-studio/components/header/ProjectHeader.jsx:76]: hmm, inlining the ternary with simple functional components seems like a nice pattern. I'll play with this!

File: apps/src/code-studio/components/header/ProjectImport.jsx
Comment (Hamms) [reply apps/src/code-studio/components/header/ProjectImport.jsx:18]: Yeah, agreed that this is super gross. I don't think it's any more gross here than it was in the header code before, but it's definitely no better, either. I added [a Jira card](https://codedotorg.atlassian.net/browse/FND-473) to track finding a better place/way to do this.

File: apps/src/code-studio/components/header/ProjectHeader.jsx
Comment (Hamms) [reply apps/src/code-studio/components/header/ProjectHeader.jsx:76]: I am going to open a separate PR for this

File: apps/src/code-studio/components/header/ProjectHeader.jsx
Comment (davidsbailey) [reply apps/src/code-studio/components/header/ProjectHeader.jsx:55]: I'm fine with punting on this, but I do think we should fix our UI tests so that they work with controlled react components. My understanding is that jquery and react use the same browser events, so it's usually just a question of which event. for example a controlled component might listen for the input event and not the change event. Next time we run into this, let's try to crack the UI test problem. I'd be happy to get pulled in on this when it comes up next.

File: apps/src/code-studio/components/header/ProjectImport.jsx
Comment (davidsbailey) [reply apps/src/code-studio/components/header/ProjectImport.jsx:18]: 100% agree that it's no more gross here than it was before, and that this is a good incremental step!

**Review by davidsbailey — APPROVED** [review]

File: apps/src/code-studio/components/header/ProjectHeader.jsx
Comment (Hamms) [reply apps/src/code-studio/components/header/ProjectHeader.jsx:55]: Hmm, my understanding was that jQuery doesn't actually trigger a real DOM event, but a simulated one that can be picked up by listeners on its own internal event handler.

See https://jsbin.com/pesanowoqe/1/edit?js,console,output for example

> Next time we run into this, let's try to crack the UI test problem. I'd be happy to get pulled in on this when it comes up next.

:+1: and likewise!

**Pull request merged**
